src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt,
+    str::FromStr,
+};
+
+use crate::instruction::{Gate, Qubit};
+
+use self::wire::Wire;
+use super::{LatexGenError, RenderCommand, RenderSettings};
+
+pub(crate) mod wire;
+
+/// Gates written in shorthand notation, i.e. composite form, that may be
+/// decomposed into modifiers and single gate instructions, i.e. canonical form.
+#[derive(Clone, Debug, strum::EnumString, derive_more::Display, PartialEq, Eq, Hash)]
+#[strum(serialize_all = "UPPERCASE")]
+enum CompositeGate {
+    /// `CNOT` is `CONTROLLED X`
+    #[display(fmt = "X")]
+    Cnot,
+    /// `CCNOT` is `CONTROLLED CONTROLLED X`
+    #[display(fmt = "X")]
+    Ccnot,
+    /// `CPHASE` is `CONTROLLED PHASE`
+    #[display(fmt = "PHASE")]
+    Cphase,
+    /// `CZ` is `CONTROLLED Z`
+    #[display(fmt = "Z")]
+    Cz,
+}
```

Comment:
@kalzoo This commit [Use regex for canonical gate decomposition and handle unimplemented X…](https://github.com/rigetti/quil-rs/pull/145/commits/68cc16eeec162a13ac3246a8cfa5fd1978c78546) removes `CanonicalGate` as an enum, and now decomposes gates into canonical form in `apply_gate`. It also handles the unimplemented `XY` gate. Since `XY` is a built-in gate, I think the best way to handle it is to look for it explicitly since any other gate of the same form could be a `DEFGATE`. So I am handling this in the same block where I look for multi-qubit gates, and explicitly checking if the canonical gate name is XY. If it is, it returns an `UnsupportedGate` error.

Reply:
Using `CONTROLLED` count to determine if a gate is a control/target gate. See [Use the generic gate and counted modifiers, and catch unsupported gat…](https://github.com/rigetti/quil-rs/pull/145/commits/755883df424663d2c49a222c9d2d90b3abe73c26)